File: src/main/java/org/veupathdb/lib/blast/PSIBlast.java
package org.veupathdb.lib.blast;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonSetter;
import org.veupathdb.lib.blast.consts.Flag;
import org.veupathdb.lib.blast.field.Location;
import org.veupathdb.lib.blast.field.Seg;

@JsonInclude(JsonInclude.Include.NON_NULL)
public class PSIBlast extends IPGListBlast
{
  private Long wordSize;

  @JsonGetter(Flag.WordSize)
  public Long getWordSize() {
    return wordSize;
  }

  @JsonSetter(Flag.WordSize)
  public void setWordSize(Long wordSize) {
    this.wordSize = wordSize;
  }

  // ------------------------------------------------------------------------------------------ //

  private Integer gapOpen;

  @JsonGetter(Flag.GapOpen)
  public Integer getGapOpen() {
    return gapOpen;
  }

  @JsonSetter(Flag.GapOpen)
  public void setGapOpen(Integer gapOpen) {
    this.gapOpen = gapOpen;
  }

  // ------------------------------------------------------------------------------------------ //

  private Integer gapExtend;

  @JsonGetter(Flag.GapExtend)
  public Integer getGapExtend() {
    return gapExtend;
  }

  @JsonSetter(Flag.GapExtend)
  public void setGapExtend(Integer gapExtend) {
    this.gapExtend = gapExtend;
  }

  // ------------------------------------------------------------------------------------------ //

  private String matrix;

  @JsonGetter(Flag.Matrix)
  public String getMatrix() {
    return matrix;
  }

  @JsonSetter(Flag.Matrix)
  public void setMatrix(String matrix) {
    this.matrix = matrix;
  }

  // ------------------------------------------------------------------------------------------ //

  private Double threshold;

  @JsonGetter(Flag.Threshold)
  public Double getThreshold() {
    return threshold;
  }

  @JsonSetter(Flag.Threshold)
  public void setThreshold(Double threshold) {
    this.threshold = threshold;
  }

  // ------------------------------------------------------------------------------------------ //

  private String compBasedStats;

  @JsonGetter(Flag.CompBasedStats)
  public String getCompBasedStats() {
    return compBasedStats;
  }

  @JsonSetter(Flag.CompBasedStats)
  public void setCompBasedStats(String compBasedStats) {
    this.compBasedStats = compBasedStats;
  }

  // ------------------------------------------------------------------------------------------ //

  private String subjectFile;

  @JsonGetter(Flag.SubjectFile)
  public String getSubjectFile() {
    return subjectFile;
  }

  @JsonSetter(Flag.SubjectFile)
  public void setSubjectFile(String subjectFile) {
    this.subjectFile = subjectFile;
  }

  // ------------------------------------------------------------------------------------------ //

  private Location subjectLocation;

  @JsonGetter(Flag.SubjectLocation)
  public Location getSubjectLocation() {
    return subjectLocation;
  }

  @JsonSetter(Flag.SubjectLocation)
  public void setSubjectLocation(Location subjectLocation) {
    this.subjectLocation = subjectLocation;
  }

  // ------------------------------------------------------------------------------------------ //

  private Seg seg;

  @JsonGetter(Flag.Seg)
  public Seg getSeg() {
    return seg;
  }

  @JsonSetter(Flag.Seg)
  public void setSeg(Seg seg) {
    this.seg = seg;
  }

  // ------------------------------------------------------------------------------------------ //

  private Long cullingLimit;

  @JsonGetter(Flag.CullingLimit)
  public Long getCullingLimit() {
    return cullingLimit;
  }

  @JsonSetter(Flag.CullingLimit)
  public void setCullingLimit(Long cullingLimit) {
    this.cullingLimit = cullingLimit;
  }

  // ------------------------------------------------------------------------------------------ //

  private Boolean sumStats;

  @JsonGetter(Flag.SumStats)
  public Boolean getSumStats() {
    return sumStats;
  }

  @JsonSetter(Flag.SumStats)
  public void setSumStats(Boolean sumStats) {
    this.sumStats = sumStats;
  }

  // ------------------------------------------------------------------------------------------ //

  private Double extensionDropoffPrelimGapped;

  @JsonGetter(Flag.ExtensionDropoffPrelimGapped)
  public Double getExtensionDropoffPrelimGapped() {
    return extensionDropoffPrelimGapped;
  }

  @JsonSetter(Flag.ExtensionDropoffPrelimGapped)
  public void setExtensionDropoffPrelimGapped(Double extensionDropoffPrelimGapped) {
    this.extensionDropoffPrelimGapped = extensionDropoffPrelimGapped;
  }

  // ------------------------------------------------------------------------------------------ //

  private Double extensionDropoffFinalGapped;

  @JsonGetter(Flag.ExtensionDropoffFinalGapped)
  public Double getExtensionDropoffFinalGapped() {
    return extensionDropoffFinalGapped;
  }

  @JsonSetter(Flag.ExtensionDropoffFinalGapped)
  public void setExtensionDropoffFinalGapped(Double extensionDropoffFinalGapped) {
    this.extensionDropoffFinalGapped = extensionDropoffFinalGapped;
  }

  // ------------------------------------------------------------------------------------------ //

  private Double gapTrigger;

  @JsonGetter(Flag.GapTrigger)
  public Double getGapTrigger() {
    return gapTrigger;
  }

  @JsonSetter(Flag.GapTrigger)
  public void setGapTrigger(Double gapTrigger) {
    this.gapTrigger = gapTrigger;
  }

  // ------------------------------------------------------------------------------------------ //

  private Short numThreads;

  @JsonGetter(Flag.NumThreads)
  public Short getNumThreads() {
    return numThreads;
  }

  @JsonSetter(Flag.NumThreads)
  public void setNumThreads(Short numThreads) {
    this.numThreads = numThreads;
  }

  // ------------------------------------------------------------------------------------------ //

  private Boolean useSmithWatermanTraceback;

  @JsonGetter(Flag.UseSmithWatermanTraceback)
  public Boolean getUseSmithWatermanTraceback() {
    return useSmithWatermanTraceback;
  }

  @JsonSetter(Flag.UseSmithWatermanTraceback)
  public void setUseSmithWatermanTraceback(Boolean useSmithWatermanTraceback) {
    this.useSmithWatermanTraceback = useSmithWatermanTraceback;
  }

  // ------------------------------------------------------------------------------------------ //

  private Double bestHitOverhang;

  @JsonGetter(Flag.BestHitOverhang)
  public Double getBestHitOverhang() {
    return bestHitOverhang;
  }

  @JsonSetter(Flag.BestHitOverhang)
  public void setBestHitOverhang(Double bestHitOverhang) {
    this.bestHitOverhang = bestHitOverhang;
  }

  // ------------------------------------------------------------------------------------------ //

  private Double bestHitScoreEdge;

  @JsonGetter(Flag.BestHitScoreEdge)
  public Double getBestHitScoreEdge() {
    return bestHitScoreEdge;
  }

  @JsonSetter(Flag.BestHitScoreEdge)
  public void setBestHitScoreEdge(Double bestHitScoreEdge) {
    this.bestHitScoreEdge = bestHitScoreEdge;
  }

  // ------------------------------------------------------------------------------------------ //

  private Boolean subjectBestHit;

  @JsonGetter(Flag.SubjectBestHit)
  public Boolean getSubjectBestHit() {
    return subjectBestHit;
  }

  @JsonSetter(Flag.SubjectBestHit)
  public void setSubjectBestHit(Boolean subjectBestHit) {
    this.subjectBestHit = subjectBestHit;
  }

  // ------------------------------------------------------------------------------------------ //

  private Long numIterations;

  @JsonGetter(Flag.NumIterations)
  public Long getNumIterations() {
    return numIterations;
  }

  @JsonSetter(Flag.NumIterations)
  public void setNumIterations(Long numIterations) {
    this.numIterations = numIterations;
  }

  // ------------------------------------------------------------------------------------------ //

  private String outPSSMFile;

  @JsonGetter(Flag.OutPSSMFile)
  public String getOutPSSMFile() {
    return outPSSMFile;
  }

  @JsonSetter(Flag.OutPSSMFile)
  public void setOutPSSMFile(String outPSSMFile) {
    this.outPSSMFile = outPSSMFile;
  }

  // ------------------------------------------------------------------------------------------ //

  private String outASCIIPSSMFile;

  @JsonGetter(Flag.OutASCIIPSSMFile)
  public String getOutASCIIPSSMFile() {
    return outASCIIPSSMFile;
  }

  @JsonSetter(Flag.OutASCIIPSSMFile)
  public void setOutASCIIPSSMFile(String outASCIIPSSMFile) {
    this.outASCIIPSSMFile = outASCIIPSSMFile;
  }

  // ------------------------------------------------------------------------------------------ //

  private Boolean savePSSMAfterLastRound;

  @JsonGetter(Flag.SavePSSMAfterLastRound)
  public Boolean getSavePSSMAfterLastRound() {
    return savePSSMAfterLastRound;
  }

  @JsonSetter(Flag.SavePSSMAfterLastRound)
  public void setSavePSSMAfterLastRound(Boolean savePSSMAfterLastRound) {
    this.savePSSMAfterLastRound = savePSSMAfterLastRound;
  }

  // ------------------------------------------------------------------------------------------ //

  private Boolean saveEachPSSM;

  @JsonGetter(Flag.SaveEachPSSM)
  public Boolean getSaveEachPSSM() {
    return saveEachPSSM;
  }

  @JsonSetter(Flag.SaveEachPSSM)
  public void setSaveEachPSSM(Boolean saveEachPSSM) {
    this.saveEachPSSM = saveEachPSSM;
  }

  // ------------------------------------------------------------------------------------------ //

  private String inMSAFile;

  @JsonGetter(Flag.InMSAFile)
  public String getInMSAFile() {
    return inMSAFile;
  }

  @JsonSetter(Flag.InMSAFile)
  public void setInMSAFile(String inMSAFile) {
    this.inMSAFile = inMSAFile;
  }

  // ------------------------------------------------------------------------------------------ //

  private Long msaMasterIDX;

  @JsonGetter(Flag.MSAMasterIDX)
  public Long getMSAMasterIDX() {
    return msaMasterIDX;
  }

  @JsonSetter(Flag.MSAMasterIDX)
  public void setMSAMasterIDX(Long msaMasterIDX) {
    this.msaMasterIDX = msaMasterIDX;
  }

  // ------------------------------------------------------------------------------------------ //

  private Boolean ignoreMSAMaster;

  @JsonGetter(Flag.IgnoreMSAMaster)
  public Boolean getIgnoreMSAMaster() {
    return ignoreMSAMaster;
  }

  @JsonSetter(Flag.IgnoreMSAMaster)
  public void setIgnoreMSAMaster(Boolean ignoreMSAMaster) {
    this.ignoreMSAMaster = ignoreMSAMaster;
  }

  // ------------------------------------------------------------------------------------------ //

  private String inPSSMFile;

  @JsonGetter(Flag.InPSSMFile)
  public String getInPSSMFile() {
    return inPSSMFile;
  }

  @JsonSetter(Flag.InPSSMFile)
  public void setInPSSMFile(String inPSSMFile) {
    this.inPSSMFile = inPSSMFile;
  }

  // ------------------------------------------------------------------------------------------ //

  private Integer pseudocount;

  @JsonGetter(Flag.Pseudocount)
  public Integer getPseudocount() {
    return pseudocount;
  }

  @JsonSetter(Flag.Pseudocount)
  public void setPseudocount(Integer pseudocount) {
    this.pseudocount = pseudocount;
  }

  // ------------------------------------------------------------------------------------------ //

  private Double inclusionEThreshold;

  @JsonGetter(Flag.InclusionEThreshold)
  public Double getInclusionEThreshold() {
    return inclusionEThreshold;
  }

  @JsonSetter(Flag.InclusionEThreshold)
  public void setInclusionEThreshold(Double inclusionEThreshold) {
    this.inclusionEThreshold = inclusionEThreshold;
  }

  // ------------------------------------------------------------------------------------------ //

  private String phiPatternFile;

  @JsonGetter(Flag.PhiPatternFile)
  public String getPhiPatternFile() {
    return phiPatternFile;
  }

  @JsonSetter(Flag.PhiPatternFile)
  public void setPhiPatternFile(String phiPatternFile) {
    this.phiPatternFile = phiPatternFile;
  }
}

File: src/main/java/org/veupathdb/lib/blast/PSIBlast.kt
package org.veupathdb.lib.blast

import com.fasterxml.jackson.annotation.JsonCreator
import org.veupathdb.lib.blast.consts.Flag
import org.veupathdb.lib.blast.field.*
import org.veupathdb.lib.blast.util.JSONObjectDecoder
import org.veupathdb.lib.blast.util.JSONObjectEncoder

open class PSIBlast(
  shortHelp: Boolean? = null,
  longHelp: Boolean? = null,
  version: Boolean? = null,
  outFormat: OutFormat? = null,
  showGIs: Boolean? = null,
  numDescriptions: NumDescriptions? = null,
  numAlignments: NumAlignments? = null,
  lineLength: LineLength? = null,
  html: Boolean? = null,
  sortHits: HitSorting? = null,
  sortHSPs: HSPSorting? = null,
  maxTargetSequences: MaxTargetSeqs? = null,
  outFile: OutFile? = null,
  parseDefLines: Boolean? = null,
  queryFile: QueryFile? = null,
  queryLocation: Location? = null,
  dbFile: String? = null,
  expectValue: ExpectValue? = null,
  softMasking: Boolean? = null,
  lowercaseMasking: Boolean? = null,
  entrezQuery: String? = null,
  queryCoverageHSPPercent: Double? = null,
  maxHSPs: Long? = null,
  dbSize: Byte? = null,
  searchSpace: Short? = null,
  importSearchStrategy: String? = null,
  exportSearchStrategy: String? = null,
  extensionDropoffUngapped: Double? = null,
  windowSize: Long? = null,
  remote: Boolean? = null,
  giList: String? = null,
  sequenceIDList: String? = null,
  negativeGIList: String? = null,
  negativeSequenceIDList: String? = null,
  taxIDs: List<String>? = null,
  negativeTaxIDs: List<String>? = null,
  taxIDList: String? = null,
  negativeTaxIDList: String? = null,
  ipgList: String? = null,
  negativeIPGList: String? = null,
  var wordSize: Long? = null,
  var gapOpen: Int? = null,
  var gapExtend: Int? = null,
  var matrix: ScoringMatrix? = null,
  var threshold: Double? = null,
  var compBasedStats: CompBasedStatsLong? = null,
  var subjectFile: String? = null,
  var subjectLocation: Location? = null,
  var seg: Seg? = null,
  var cullingLimit: Long? = null,
  var sumStats: Boolean? = null,
  var extensionDropoffPrelimGapped: Double? = null,
  var extensionDropoffFinalGapped: Double? = null,
  var gapTrigger: Double? = null,
  var numThreads: Short? = null,
  var useSmithWatermanTraceback: Boolean? = null,
  var bestHitOverhang: Double? = null,
  var bestHitScoreEdge: Double? = null,
  var subjectBestHit: Boolean? = null,
  var numIterations: Long? = null,
  var outPSSMFile: String? = null,
  var outASCIIPSSMFile: String? = null,
  var savePSSMAfterLastRound: Boolean? = null,
  var saveEachPSSM: Boolean? = null,
  var inMSAFile: String? = null,
  var mSAMasterIDX: Long? = null,
  var ignoreMSAMaster: Boolean? = null,
  var inPSSMFile: String? = null,
  var pseudocount: Int? = null,
  var inclusionEThreshold: Double? = null,
  var phiPatternFile: String? = null,
) : BlastWithIPGList(
  shortHelp,
  longHelp,
  version,
  outFormat,
  showGIs,
  numDescriptions,
  numAlignments,
  lineLength,
  html,
  sortHits,
  sortHSPs,
  maxTargetSequences,
  outFile,
  parseDefLines,
  queryFile,
  queryLocation,
  dbFile,
  expectValue,
  softMasking,
  lowercaseMasking,
  entrezQuery,
  queryCoverageHSPPercent,
  maxHSPs,
  dbSize,
  searchSpace,
  importSearchStrategy,
  exportSearchStrategy,
  extensionDropoffUngapped,
  windowSize,
  remote,
  giList,
  sequenceIDList,
  negativeGIList,
  negativeSequenceIDList,
  taxIDs,
  negativeTaxIDs,
  taxIDList,
  negativeTaxIDList,
  ipgList,
  negativeIPGList
), BlastQueryConfig {
  override val tool = BlastTool.PSIBlast

  @JsonCreator
  constructor(js: JSONObjectDecoder) : this() {
    decodeJSON(js)
  }

  override fun toJSON(includeTool: Boolean): JSONObjectEncoder {
    val js: JSONObjectEncoder = super<BlastWithIPGList>.toJSON(includeTool)
    js.encode(Flag.WordSize, wordSize)
    js.encode(Flag.GapOpen, gapOpen)
    js.encode(Flag.GapExtend, gapExtend)
    js.encode(Flag.Matrix, matrix)
    js.encode(Flag.Threshold, threshold)
    js.encode(Flag.CompBasedStats, compBasedStats)
    js.encode(Flag.SubjectFile, subjectFile)
    js.encode(Flag.SubjectLocation, subjectLocation)
    js.encode(Flag.Seg, seg)
    js.encode(Flag.CullingLimit, cullingLimit)
    js.encode(Flag.SumStats, sumStats)
    js.encode(Flag.ExtensionDropoffPrelimGapped, extensionDropoffPrelimGapped)
    js.encode(Flag.ExtensionDropoffFinalGapped, extensionDropoffFinalGapped)
    js.encode(Flag.GapTrigger, gapTrigger)
    js.encode(Flag.NumThreads, numThreads)
    js.encode(Flag.UseSmithWatermanTraceback, useSmithWatermanTraceback)
    js.encode(Flag.BestHitOverhang, bestHitOverhang)
    js.encode(Flag.BestHitScoreEdge, bestHitScoreEdge)
    js.encode(Flag.SubjectBestHit, subjectBestHit)
    js.encode(Flag.NumIterations, numIterations)
    js.encode(Flag.OutPSSMFile, outPSSMFile)
    js.encode(Flag.OutASCIIPSSMFile, outASCIIPSSMFile)
    js.encode(Flag.SavePSSMAfterLastRound, savePSSMAfterLastRound)
    js.encode(Flag.SaveEachPSSM, saveEachPSSM)
    js.encode(Flag.InMSAFile, inMSAFile)
    js.encode(Flag.MSAMasterIDX, mSAMasterIDX)
    js.encode(Flag.IgnoreMSAMaster, ignoreMSAMaster)
    js.encode(Flag.InPSSMFile, inPSSMFile)
    js.encode(Flag.Pseudocount, pseudocount)
    js.encode(Flag.InclusionEThreshold, inclusionEThreshold)
    js.encode(Flag.PhiPatternFile, phiPatternFile)
    return js
  }

  override fun decodeJSON(node: JSONObjectDecoder) {
    super<BlastWithIPGList>.decodeJSON(node)

    node.decodeLong(Flag.WordSize) { wordSize = it }
    node.decodeInt(Flag.GapOpen) { gapOpen = it }
    node.decodeInt(Flag.GapExtend) { gapExtend = it }
    node.decodeJSON(Flag.Matrix) { matrix = ScoringMatrix.fromJSON(it) }
    node.decodeDouble(Flag.Threshold) { threshold = it }
    node.decodeJSON(Flag.CompBasedStats) { compBasedStats = CompBasedStatsLong.fromJSON(it) }
    node.decodeString(Flag.SubjectFile) { subjectFile = it }
    node.decodeJSON(Flag.SubjectLocation) { subjectLocation = Location(it)}
    node.decodeJSON(Flag.Seg) { seg = Seg.fromJSON(it) }
    node.decodeLong(Flag.CullingLimit) { cullingLimit = it }
    node.decodeBool(Flag.SumStats) { sumStats = it }
    node.decodeDouble(Flag.ExtensionDropoffPrelimGapped) { extensionDropoffPrelimGapped = it }
    node.decodeDouble(Flag.ExtensionDropoffFinalGapped) { extensionDropoffFinalGapped = it }
    node.decodeDouble(Flag.GapTrigger) { gapTrigger = it }
    node.decodeShort(Flag.NumThreads) { numThreads = it }
    node.decodeBool(Flag.UseSmithWatermanTraceback) { useSmithWatermanTraceback = it }
    node.decodeDouble(Flag.BestHitOverhang) { bestHitOverhang = it }
    node.decodeDouble(Flag.BestHitScoreEdge) { bestHitScoreEdge = it }
    node.decodeBool(Flag.SubjectBestHit) { subjectBestHit = it }
    node.decodeLong(Flag.NumIterations) { numIterations = it }
    node.decodeString(Flag.OutPSSMFile) { outPSSMFile = it }
    node.decodeString(Flag.OutASCIIPSSMFile) { outASCIIPSSMFile = it }
    node.decodeBool(Flag.SavePSSMAfterLastRound) { savePSSMAfterLastRound = it }
    node.decodeBool(Flag.SaveEachPSSM) { saveEachPSSM = it }
    node.decodeString(Flag.InMSAFile) { inMSAFile = it }
    node.decodeLong(Flag.MSAMasterIDX) { mSAMasterIDX = it }
    node.decodeBool(Flag.IgnoreMSAMaster) { ignoreMSAMaster = it }
    node.decodeString(Flag.InPSSMFile) { inPSSMFile = it }
    node.decodeInt(Flag.Pseudocount) { pseudocount = it }
    node.decodeDouble(Flag.InclusionEThreshold) { inclusionEThreshold = it }
    node.decodeString(Flag.PhiPatternFile) { phiPatternFile = it }
  }
}